### DISCUSSION TITLE
Slice snapping

### DISCUSSION THREAD

**Discussion by fcurow** [firemodels/fds#13417] [General]
To whom it may concern,

Was there ever a version of FDS that snapped slices (SLCF) from their requested location to the mesh location, similar to what is done with obstructions?

Best Regards,
Sal

**Comment by rmcdermo** [comment]
That is still the case.  Now you can also specify `CELL_CENTERED=T`, but slices either are interpolated to vertices or left at cell center.

**Reply by fcurow** [reply]
Thank you. So, visually in smokeview, the values shown are discretized down to either cell centered or vertex values? There is no further interpolation? The visual representation of the sclf in smokeview corresponds to the closest cell center or vertex value?

**Comment by gforney** [comment]
for node centered slices as randy says smokeview draw data at vertices, the
graphic card then interpolates the colors between vertices.  for cell
centered slices smokeview draws the same color for the entire cell,
centered at the cell center

On Mon, Sep 9, 2024 at 3:14 PM fcurow ***@***.***> wrote:

> Thank you. So, visually in smokeview, the values shown are discretized
> down to either cell centered or vertex values? There is no further
> interpolation? The visual representation of the sclf in smokeview
> corresponds to the closest cell center or vertex value?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/firemodels/fds/discussions/13417#discussioncomment-10594866>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC6UCRVAJARM563NTU3QK2TZVXXRFAVCNFSM6AAAAABN5DVQL2VHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTANJZGQ4DMNQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Glenn Forney

**Reply by fcurow** [reply]
Thank you. If my slice does not physically go through a node nor cell center. CELL_CENTERED=T would results in the same color for the portion of the slice within that cell and correspond to the value at that cell center and if CELL_CENTERED=F then, as you write, there will be interpolation between vertices. Will:
1) the slice location physically move in either instance?
2) when the interpolation occurs, the level of interpolation is a function of the available pixilation? That interpolation would be based on the 4 closest vertices weighted by distance?

**Comment by gforney** [comment]
the slice location does not move, through cell centers for cell centered
slice files and along cell face for node centered slice files.  smokeview
draws triangles not rectangles.  it creates a vertex at the center of the
cell face with value the average of the 4 values at rectangle vertices.  it
draws 4 triangles for each rectangle, each triangle uses the center vertex
and 2 adjacent vertices of the rectangle


On Mon, Sep 9, 2024 at 3:44 PM fcurow ***@***.***> wrote:

> Thank you. If my slice does not physically go through a node nor cell
> center. CELL_CENTERED=T would results in the same color for the portion of
> the slice within that cell and correspond to the value at that cell center
> and if CELL_CENTERED=F then, as you write, there will be interpolation
> between vertices. Will:
>
>    1. the slice location physically move in either instance?
>    2. when the interpolation occurs, the level of interpolation is a
>    function of the available pixilation? That interpolation would be based on
>    the 4 closest vertices weighted by distance?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/firemodels/fds/discussions/13417#discussioncomment-10595075>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC6UCRT4C24N322QHI5KLSDZVX3AVAVCNFSM6AAAAABN5DVQL2VHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTANJZGUYDONI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Glenn Forney

**Reply by fcurow** [reply]
Thank you, if I place my slice in-between nodes and cell centers and the slice location is fixed. how does the program assign values in this case?

Thank you, so it discretizes into 4 sections per cell face when interpolating.

**Reply by mcgratta** [reply]
For the default slice, values are linearly interpolated at cell corners using the nearest 8 values that live either at cell centers or cell faces. For a `CELL_CENTERED=T` case, there is no interpolation, You get the computed value at either the cell center or face.

**Reply by drjfloyd** [reply]
There is no interpolation to the exact slice location. The SLCF inputs just point to what grid cells to ouptut data for.  If you had this mesh:

&MESH IJK=10,10,10,XB-0,1,0,1,0,1/ 

Then these slice files would all have the same data as they all point to the first plane of grid cells in the x-direction:
&SLCF PBX=0.01,QUANTITY='TEMPERATURE'/
&SLCF PBX=0.02,QUANTITY='TEMPERATURE'/
&SLCF PBX=0.09,QUANTITY='TEMPERATURE'/

**Reply by fcurow** [reply]
Thank you all for taking the time, I have a much better understanding now. I think I have one final question. If my understanding is correct, if I have a slice that goes through the cell center, but CELL_CENTERED=F then the slcf will point to the vertices/node values, but if it is in the center of the cell it is then equal distant to vertices on one side versus the other side of the slice location, is there a convention to which vertices the slice files point?

**Reply by mcgratta** [reply]
You would think that 15.5 would round up to 16. But on the computer, there is no such thing as 15.5. It might be 15.49999999999999 or 15.5000000001. So if you really want a slice to live at  a particular place, do not split the cell "exactly".

**Reply by fcurow** [reply]
Thank you, but what if I do split the cell in the middle i.e. 15.5000000000, and the cell size is 1 unit. Does the slice file point to vertices at 16 or 15?

**Reply by rmcdermo** [reply]
Try it on a simple test case.  It will depend on the system.

**Comment by mcgratta** [comment]
My guess is 16, but one can never be sure because the numerical representation of 15.5 is operating system dependent.

**Reply by drjfloyd** [reply]
If you want to understand how real numbers are stored here is an example using IEEE 754. There are other standard methods.

https://www.computerengineeringconcepts.org/2.7-Real-Number-Representation

**Reply by fcurow** [reply]
Wow, thanks guys!
